efs_facts: improve performance by reducing the number of api calls

julienvey opened the pull request:
##### SUMMARY
<!--- Describe the change, including rationale and design decisions -->
Related to #36506 and comment https://github.com/ansible/ansible/issues/36506#issuecomment-367343565
<!---
If you are fixing an existing issue, please include "Fixes #nnn" in your
commit message and your description; but you should still explain what
the change does.
-->


##### ISSUE TYPE
<!--- Pick one below and delete the rest: -->
 - Bugfix Pull Request

##### COMPONENT NAME
<!--- Name of the module, plugin, module or task -->
efs_facts

##### ANSIBLE VERSION
<!--- Paste verbatim output from "ansible --version" between quotes below -->
```
2.6.0
```


##### ADDITIONAL INFORMATION
<!---
Include additional information to help people understand the change here.
For bugs that don't have a linked bug report, a step-by-step reproduction
of the problem is helpful.
  -->
I have been profiling the tasks before and after the changes and the results are a lot more consistent and fast with this new code

julienvey committed:
efs_facts: improve performance by reducing the number of api calls

s-hertel commented on lib/ansible/modules/cloud/amazon/efs_facts.py:
```diff
@@ -348,10 +361,14 @@ def main():
         targets = [(item, prefix_to_attr(item)) for item in targets]
         file_systems_info = [item for item in file_systems_info if has_targets(item['mount_targets'], targets)]
 
+    file_systems_info = connection.get_security_groups_data(file_systems_info)
```
Right now a task like:
```
    - name: get efs facts using targets
      efs_facts:
        profile: shertel
        region: us-east-1
        targets:
          - 'sg-5e27db2f'
```
doesn't return anything because file_systems_info is being overwritten by the list comprehension above. This line should go above `if targets:`. Looks good to merge besides that.

julienvey replied:
Good catch

julienvey commented:
@s-hertel I've updated the code to fix the issue you raised in your comment. I've also added a bunch of integration tests that should cover most of the case.

However, this PR will need a CI upgrade, shippable is failing right now. I've added a policy file which describes all the necessary permissions to run the tests.

ryansydnor reviewed (approved):
thanks for adding tests!

julienvey commented:
@s-hertel Do you have an update on the ci update needed for this PR ? Thanks

s-hertel commented:
This should have the necessary permissions, I'm trying to determine why it is still failing.

s-hertel committed:
Remove efs_facts tests from running in CI

s-hertel committed:
fix aliases

s-hertel commented:
Now that there are alias labels to add that keep the tests from running in CI I'm just going to merge this and open a backport PR. The problem is essentially that roles cannot create mount targets (try running the tests with an IAM user and then an assumed role with the exact same policy), and I can't find a way to do it. Anyone merging EFS things will need to run the tests locally first.

I can't find this behavior documented anywhere, so it's very possible I'm just not trying right. If someone can shed some light, or if we can get these running in CI at some point it would be great, but this shouldn't be waiting on it. Thanks for the extreme patience @julienvey.

s-hertel commented:
bot_status
